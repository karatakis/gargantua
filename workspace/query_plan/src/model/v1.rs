use std::collections::HashSet;

use async_graphql_parser::types::{self as Q};

use crate::error::Error;

pub struct TraitSet<A>(HashSet<A>);

pub struct Node<A, T, P> {
    pub data: A,
    pub children: Vec<Node<A, T, P>>,
    pub traits: TraitSet<T>,
    pub primary: P,
}

impl<A, T> Node<A, T, ()> {
    pub fn try_from(sel: Q::SelectionSet) -> Result<Self, Error> {
        todo!()
    }
}
