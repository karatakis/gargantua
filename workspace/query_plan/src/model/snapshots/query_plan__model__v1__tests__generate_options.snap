---
source: workspace/query_plan/src/model/v1.rs
expression: options
---
[
    Node {
        traits: Only(
            "Location",
        ),
        data: "test",
        children: [
            Node {
                traits: Only(
                    "Reviews",
                ),
                data: "p",
                children: [],
            },
            Node {
                traits: Only(
                    "Location",
                ),
                data: "r",
                children: [],
            },
        ],
    },
    Node {
        traits: Only(
            "Reviews",
        ),
        data: "test",
        children: [
            Node {
                traits: Only(
                    "Reviews",
                ),
                data: "p",
                children: [],
            },
            Node {
                traits: Only(
                    "Location",
                ),
                data: "r",
                children: [],
            },
        ],
    },
    Node {
        traits: All(
            [
                "Location",
                "Reviews",
            ],
        ),
        data: "test",
        children: [
            Node {
                traits: Only(
                    "Reviews",
                ),
                data: "p",
                children: [],
            },
            Node {
                traits: Only(
                    "Location",
                ),
                data: "r",
                children: [],
            },
        ],
    },
    Node {
        traits: Only(
            "Location",
        ),
        data: "test",
        children: [
            Node {
                traits: Only(
                    "Reviews",
                ),
                data: "p",
                children: [],
            },
            Node {
                traits: Only(
                    "Reviews",
                ),
                data: "r",
                children: [],
            },
        ],
    },
    Node {
        traits: Only(
            "Reviews",
        ),
        data: "test",
        children: [
            Node {
                traits: Only(
                    "Reviews",
                ),
                data: "p",
                children: [],
            },
            Node {
                traits: Only(
                    "Reviews",
                ),
                data: "r",
                children: [],
            },
        ],
    },
    Node {
        traits: All(
            [
                "Location",
                "Reviews",
            ],
        ),
        data: "test",
        children: [
            Node {
                traits: Only(
                    "Reviews",
                ),
                data: "p",
                children: [],
            },
            Node {
                traits: Only(
                    "Reviews",
                ),
                data: "r",
                children: [],
            },
        ],
    },
    Node {
        traits: Only(
            "Location",
        ),
        data: "test",
        children: [
            Node {
                traits: Only(
                    "Reviews",
                ),
                data: "p",
                children: [],
            },
            Node {
                traits: All(
                    [
                        "Location",
                        "Reviews",
                    ],
                ),
                data: "r",
                children: [],
            },
        ],
    },
    Node {
        traits: Only(
            "Reviews",
        ),
        data: "test",
        children: [
            Node {
                traits: Only(
                    "Reviews",
                ),
                data: "p",
                children: [],
            },
            Node {
                traits: All(
                    [
                        "Location",
                        "Reviews",
                    ],
                ),
                data: "r",
                children: [],
            },
        ],
    },
    Node {
        traits: All(
            [
                "Location",
                "Reviews",
            ],
        ),
        data: "test",
        children: [
            Node {
                traits: Only(
                    "Reviews",
                ),
                data: "p",
                children: [],
            },
            Node {
                traits: All(
                    [
                        "Location",
                        "Reviews",
                    ],
                ),
                data: "r",
                children: [],
            },
        ],
    },
]
